go-tlsdialer: initial repository setup

DerekBum opened the pull request:
To disable SSL by default we want to transfer OpenSSLDialer and any other SSL logic to the new go-tlsdialer repository.

go-tlsdialer serves as an interlayer between go-tarantool and go-openssl. All SSL logic from go-tarantool is moved to the go-tlsdialer.

go-tlsdialer still uses tarantool connection, but also types and methods from go-openssl. This way we are removing the direct go-openssl dependency from go-tarantool, without creating a tarantool dependency in go-openssl.

Moved all SSL code from go-tarantool, some test helpers. Added `README.md`, ci workflow.

Part of https://github.com/tarantool/go-tarantool/issues/301

DerekBum commented:
For the most part, code copied here will be deleted from the go-tarantool repo. However, there is the list of the parts, that will be in both repositories:
1. `writeFlusher` interface.
2. `deadlineIO` type and its methods.
3. `parseAddress` function.
4. Some of the test helpers in the `ssl_test.go`.

oleg-jukovec commented on .github/workflows/testing.yml:
```diff
@@ -0,0 +1,79 @@
+      - name: Setup golang for the connector and tests
+        uses: actions/setup-go@v3
+        with:
+          go-version: '1.20'
```
It would be nice to test with latest version too and macOS if it possible.

oleg-jukovec commented on .golangci.yaml:
```diff
@@ -0,0 +1,30 @@
+  timeout: 3m
+
+linters:
+  disable:
+    - errcheck
```
We could do it here with the checker. 
```suggestion
```

oleg-jukovec commented on .luacheckrc:
```diff
@@ -0,0 +1,28 @@
+max_line_length = 120
+
+ignore = {
+    "212/self",   -- Unused argument <self>.
+    "411",        -- Redefining a local variable.
+    "421",        -- Shadowing a local variable.
+    "431",        -- Shadowing an upvalue.
+    "432",        -- Shadowing an upvalue argument.
```
Do we really need it?

oleg-jukovec commented on .github/workflows/testing.yml:
```diff
@@ -0,0 +1,79 @@
+        shell: bash
+        run: echo "COVERAGES=$(find . -type f -name 'module-coverage.txt' | tr -s '\n' ',' | sed 's/,$//')" >> $GITHUB_ENV
+      - name: Upload coverage to Codecov
+        uses: codecov/codecov-action@81cd2dc8148241f03f5839d295e000b8f761e378 # v3.1.0
```
Please, make sure that it works.

oleg-jukovec commented on README.md:
```diff
@@ -0,0 +1,93 @@
+# tlsdialer
+
+Package go-tlsdialer serves as an interlayer between go-tarantool and go-openssl. 
+All ssl logic from go-tarantool is moved to the go-tlsdialer.
+
+go-tlsdialer still uses tarantool connection, but also types and methods from 
+go-openssl. 
+This way we are removing the direct go-openssl dependency from go-tarantool,
+without creating a tarantool dependency in go-openssl.
```
It looks like what we want to achieve with this package, not what this package does. I suppose something like:

```The package allows to create a TLS dialer for `go-tarantool (link)`...```

oleg-jukovec commented on README.md:
```diff
@@ -0,0 +1,93 @@
+# tlsdialer
```
It would be nice to have the godoc badge:
https://github.com/tarantool/go-tarantool/blob/f02579a563687a53600f9988b7eba7ebced293c8/README.md?plain=1#L5

And the coverage badge:
https://github.com/tarantool/go-tarantool/blob/f02579a563687a53600f9988b7eba7ebced293c8/README.md?plain=1#L7

oleg-jukovec commented on README.md:
````diff
@@ -0,0 +1,93 @@
+}
+```
+To create a connection from the created dialer a `Dial` function could be used:
+```go
+dialer := tlsdialer.OpenSslDialer{
+	... // fill structure fields here.
+}
+opts := tarantool.Opts{
+    Timeout: 5 * time.Second,
+}
+
+ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
+defer cancel()
+
+conn, err := tarantool.Connect(ctx, dialer, opts)
+if err != nil {
+    fmt.Printf("Failed to create an example connection: %s", err)
+    return
+}
+
+// Use the connection.
+...
+```
````
Let's add a full example. With all imports, main function etc, like:

https://github.com/tarantool/go-tarantool/blob/f02579a563687a53600f9988b7eba7ebced293c8/README.md?plain=1#L105-L140

oleg-jukovec commented on config.lua:
I think the full file is not necessary for our tests. Please, recheck it and remove unnecessary code.

oleg-jukovec commented on config.lua:
Please, move it to `testdata/config.lua`.

https://pkg.go.dev/cmd/go/internal/test

`The go tool will ignore a directory named "testdata", making it available to hold ancillary data needed by the tests.`

oleg-jukovec commented on connection.go:
`connection.go` -> `conn.go`, like a type name.

oleg-jukovec commented on ssl_dial.go:
`ssl_dial.go` -> `dial.go`

oleg-jukovec commented on ssl_opts.go:
```diff
@@ -0,0 +1,30 @@
+package tlsdialer
+
+type sslOpts struct {
```
```suggestion
type opts struct {
```

oleg-jukovec commented on ssl_opts.go:
ssl_opts.go -> opts.go

oleg-jukovec commented on tarantool_test.go:
`tarantool_test.go` is not totally clear, let's name it `integration_test.go`

oleg-jukovec commented on tarantool_test.go:
```diff
@@ -0,0 +1,58 @@
+
+	return m.Run()
+}
+
```
Please, move all integration tests with real Tarantool to the file.

oleg-jukovec commented on testdata/sidecar/main:
Please, remove the binary.

oleg-jukovec commented on ssl_test.go:
We test the api of the Dialer here.

ssl_test.go -> dialer_test.go

oleg-jukovec reviewed (commented):
Please, split the commit into commits:
* CI
* README
* Implementation

Let's also add:
* `example_test.go` file with a short example.
* `CHANGELOG.md` with a shot description of the changes.

You could although remove all `ssl*` prefixes at now.

DerekBum committed:
ci: setup ci workflow

In this commit the ci workflow is created for the repository.

DerekBum committed:
docs: create a README file

In thes commit `README.md` and `CHANGELOG.md` files are created.
`README` shows the purpose of this repository.

DerekBum committed:
api: add `OpenSSLDialer` implementation

To disable SSL by default we want to transfer OpenSSLDialer
and any other SSL logic to the new go-tlsdialer repository.

go-tlsdialer serves as an interlayer between go-tarantool and
go-openssl. All SSL logic from go-tarantool is moved to the
go-tlsdialer.

go-tlsdialer still uses tarantool connection, but also
types and methods from go-openssl. This way we are
removing the direct go-openssl dependency from go-tarantool,
without creating a tarantool dependency in go-openssl.

Moved all SSL code from go-tarantool, some test helpers.

Part of https://github.com/tarantool/go-tarantool/issues/301

oleg-jukovec commented on README.md:
```diff
@@ -0,0 +1,121 @@
+		Tuple([]interface{}{99999, "BB"}),
+	).Get()
+	if err != nil {
+		fmt.Println("Error", err)
```
To make it the same for all example:
```suggestion
		fmt.Printlf("Error: %s", err)
```

oleg-jukovec commented on dial.go:
```diff
@@ -0,0 +1,143 @@
+	return openssl.DialContext(ctx, network, address, sslCtx.(*openssl.Ctx), 0)
+}
+
+// interface{} is a hack. It helps to avoid dependency of go-openssl in build
+// of tests with the tag 'go_tarantool_ssl_disable'.
+func sslCreateContext(sslOpts opts) (ctx interface{}, err error) {
```
We don't need the hack now.
```suggestion
func sslCreateContext(sslOpts opts) (sslCtx *openssl.Ctx}, err error) {
```

oleg-jukovec commented on dial.go:
```diff
@@ -0,0 +1,143 @@
+		return
+	}
+
+	return openssl.DialContext(ctx, network, address, sslCtx.(*openssl.Ctx), 0)
```
After the previous fix:
```suggestion
	return openssl.DialContext(ctx, network, address, sslCtx, 0)
```

oleg-jukovec commented on example_test.go:
```diff
@@ -0,0 +1,64 @@
+	// Use the connection.
+	data, err := conn.Do(tarantool.NewPingRequest()).Get()
+	if err != nil {
+		fmt.Println("Error", err)
```
The same:
```suggestion
		fmt.Printf("Error: %s", err)
```

oleg-jukovec commented on testdata/config.lua:
```diff
@@ -0,0 +1,40 @@
+box.cfg{
+    auth_type = auth_type,
+    work_dir = os.getenv("TEST_TNT_WORK_DIR"),
+    memtx_use_mvcc_engine = os.getenv("TEST_TNT_MEMTX_USE_MVCC_ENGINE") == 'true' or nil,
```
I think we don't `memtx_use_mvcc_engine` here (it is required by streams tests).

oleg-jukovec commented on testdata/sidecar/main.go:
Oh, the file about fd dialer. You could remove it with the `sidecar` folder.

oleg-jukovec commented on dialer_test.go:
```diff
@@ -0,0 +1,433 @@
+	expectedErr          string
+	expectedProtocolInfo tarantool.ProtocolInfo
+
+	// These options configure the behavior of the server.
+	isErrGreeting   bool
+	isErrId         bool
+	isIdUnsupported bool
+	isPapSha256Auth bool
+	isErrAuth       bool
+	isEmptyAuth     bool
```
Do we use all options in the tests? You could remove some of them.

DerekBum replied:
Removed `isEmptyAuth` as the only one that wasn't used in the tests.

oleg-jukovec commented on dialer_test.go:
```diff
@@ -0,0 +1,433 @@
+var (
+	opts = tarantool.Opts{
+		Timeout: 5 * time.Second,
+		//Concurrency: 32,
+		//RateLimit: 4*1024,
```
What for?

oleg-jukovec reviewed (approved):
Thank you for the patch!

DerekBum commented:
Fixed everything from the previous comments.

oleg-jukovec commented on .github/workflows/testing.yml:
```diff
@@ -0,0 +1,171 @@
+    #
+    # The main trick is described here:
+    # https://github.com/Dart-Code/Dart-Code/pull/2375
+    if: (github.event_name == 'push') ||
+      (github.event_name == 'pull_request' &&
+        github.event.pull_request.head.repo.full_name != github.repository) ||
+      (github.event_name == 'workflow_dispatch')
```
Please, copy conditions from `run-tests-ee`. It allows to run CI for pull requests from an external repositories correctly.

DerekBum replied:
Updated.

oleg-jukovec commented on deadline_io.go:
`deadline_io.go` -> `deadlineio.go`

https://stackoverflow.com/a/39625565

DerekBum replied:
Updated.

oleg-jukovec commented on deadline_io.go:
```diff
@@ -0,0 +1,31 @@
+	c  net.Conn
+}
+
+func (d *deadlineIO) Write(b []byte) (n int, err error) {
```
Let's also add a unit-test to ensure that `diadlineIO` works fine. So we need to test:

https://github.com/tarantool/go-tlsdialer/blob/927312d079c272de904d378c9ab9793d329757a4/dialer.go#L24

that the `tarantool.DialOpts` value applies successfully for the `OpenSSLDialer` and code works as expected.

DerekBum replied:
Added new test to the `integration_test.go`.

DifferentialOrange commented on .github/workflows/testing.yml:
```diff
@@ -0,0 +1,174 @@
+      fail-fast: false
+      matrix:
+        golang:
+          - '1.20'
```
Doesn't go-tarantool still supports 1.13?

DerekBum replied:
`go-tarantool v2` will also update go version to 1.20.

DifferentialOrange commented on CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 
 ### Added
 
+* `OpenSslDialer` type (#1).
```
I think it's better to have a short useful description like "`OpenSslDialer` type to use SSL transport for `tarantool/go-tarantool/v2` connection".

DerekBum replied:
Updated.

DifferentialOrange commented on conn.go:
```diff
@@ -0,0 +1,66 @@
+	"github.com/tarantool/go-tarantool/v2"
+)
+
+type tntConn struct {
```
AFAIR, we now stick to `tt` short name instead of `tnt` for `tarantool`.

DerekBum replied:
Updated.

DifferentialOrange commented on conn.go:
```diff
@@ -0,0 +1,66 @@
+
+// ProtocolInfo makes tntConn satisfy the Conn interface.
+func (c *tntConn) ProtocolInfo() tarantool.ProtocolInfo {
+	return tarantool.ProtocolInfo{}
```
What's about the actual ones?

DerekBum replied:
Actual `ProtocolInfo` is filled in the `ProtocolDialer`. Actual `Greeting` is filled in the `GreetingDialer`.

DifferentialOrange replied:
So if I want to use SSL with protocol features, I need to build my own Frankenstein from dialers?

oleg-jukovec replied:
No, you are not. We already done it for a user:

https://github.com/tarantool/go-tlsdialer/blob/94b60638fb8b3cbe3c2514e5fd90d66780095119/dialer.go#L48-L126

DerekBum replied:
No, `OpenSslDialer` builds this "Frankenstein" itself, without user knowing:
https://github.com/tarantool/go-tlsdialer/blob/94b60638fb8b3cbe3c2514e5fd90d66780095119/dialer.go#L99-L123

But if you want to build your own dialer, then yes. You need to choose what dialer-helpers to use (as example, do you need authentication for your dialer or not), and build your custom dialer from blocks.

DerekBum commented:
Updated the `README`, now it includes instructions about the application build (static and dynamic).

oleg-jukovec commented on README.md:
```diff
@@ -0,0 +1,169 @@
+
+## Application build
+
+Since tlsdialer uses SSL for connection to the Tarantool-EE, Cgo should be
```
It is more clear.
```suggestion
Since tlsdialer uses OpenSSL for connection to the Tarantool-EE, Cgo should be
```

DerekBum replied:
Updated.

oleg-jukovec commented on README.md:
````diff
@@ -0,0 +1,169 @@
+1. **Static build**. Run this command from the installation directory to configure
+   the OpenSSL:
+   ```shell
+   ./config no-shared --prefix=/tmp/openssl/ --openssldir=/usr/local/ssl/
````
Why do we need to specify `openssldir` here?
```suggestion
   ./config no-shared --prefix=/tmp/openssl/
```
I think it is not required.

DerekBum replied:
Updated.

oleg-jukovec commented on README.md:
````diff
@@ -0,0 +1,169 @@
+   ./config --prefix=/tmp/openssl/ --openssldir=/usr/local/ssl/
+   ```
+   After configuring, run this command to install and build OpenSSL:
+   ```shell
+   make install --prefix=/tmp/openssl/
+   ```
````
It will install to the prefix anyway.
```suggestion
   ```shell
   make install
   ```
```

DerekBum replied:
Updated.

oleg-jukovec commented on README.md:
````diff
@@ -0,0 +1,169 @@
+go test -v ./...
+```
+
+## OpenSslDialer
````
OpenSSL library, OpenSSLDialer.. looks fine.
```suggestion
## OpenSSLDialer
```

DerekBum replied:
Updated.

oleg-jukovec commented on README.md:
````diff
@@ -0,0 +1,169 @@
+User can create a dialer by filling the struct:
+```go
+// OpenSslDialer allows to use SSL transport for connection.
+type OpenSslDialer struct {
````
```suggestion
type OpenSSLDialer struct {
```

DerekBum replied:
Updated.

oleg-jukovec commented on dialer.go:
```diff
@@ -0,0 +1,141 @@
+}
+
+// OpenSslDialer allows to use SSL transport for connection.
+type OpenSslDialer struct {
```
```suggestion
type OpenSSLDialer struct {
```

DerekBum replied:
Updated.
